File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.py
import heapq

class Solution:
    def maxScore(self, nums1: List[int], nums2: List[int], k: int) -> int:
        z = list(zip(nums1, nums2))
        z.sort(key = lambda x: -x[-1])
        _sum = 0
        _min = float('inf')
        
        h = []
        
        for i in range(k):
            n1, n2 = z[i]
            _sum += n1
            heapq.heappush(h, (n1, n2))
            
        answer = _sum * z[k - 1][1]
        for i in range(k, len(z)):
            n1, n2 = z[i]
            sn1, sn2 = heapq.heappop(h)
            _sum -= sn1
            _sum += n1
            heapq.heappush(h, (n1, n2))
            answer = max(answer, _sum * n2)
        
        return answer
            
            
        
            
        

File: 2542-maximum-subsequence-score/2542-maximum-subsequence-score.rs
use std::{collections::BinaryHeap, cmp};

impl Solution {
    pub fn max_score(nums1: Vec<i32>, nums2: Vec<i32>, k: i32) -> i64 {
        let mut nums:Vec<(i64, i64)> = vec![];
        let mut heap: BinaryHeap<_> = BinaryHeap::new();
        
        let n = nums1.len();
        let k: usize = k as usize;
        
        for i in 0..n {
            let (n1, n2) = (nums1[i], nums2[i]);
            nums.push((n1 as i64, n2 as i64));
        }
        
        nums.sort_by(|a, b| b.1.cmp(&a.1));
        let mut sum: i64 = 0;
        
        for i in 0..k {
            let (n1, n2) = nums[i];
            heap.push(-n1);
            sum += n1;
        }
        
        let mut answer: i64 = (sum * nums[k - 1].1) as i64;
  
        for i in k..n {
            let (n1, n2) = nums[i];
            let smallest = heap.pop().unwrap();
            sum += smallest;
            sum += n1;
            heap.push(-n1);
            answer = cmp::max(answer, (sum * n2) as i64);
        }
        
        return answer
    }
}